oxidation/libparsec/crates/serialization_format/src/protocol.rs

```diff
@@ -0,0 +1,739 @@
+                                        name: nested_type.name.to_owned(),
+                                        discriminant_field: discriminant_field.to_owned(),
+                                        variants: variants.iter().map(
+                                            |variant| {
+                                                GenCmdNestedTypeVariant {
+                                                    name: variant.name.to_owned(),
+                                                    discriminant_value: variant.discriminant_value.to_owned(),
+                                                    fields: variant.fields.as_ref().map(|fields| {
+                                                        fields.iter()
+                                                        .filter_map(&convert_field)
+                                                        .collect()
+                                                    }),
+                                                }
+                                            }
+                                        ).collect()
+                                    }
+                                },
+                                (Some(fields), None, None) => {
+                                    GenCmdNestedType::Struct {
+                                        name: nested_type.name.to_owned(),
+                                        fields: fields
+                                            .iter()
+                                            .filter_map(&convert_field)
+                                            .collect(),
+                                    }
+                                }
+                                _ => {
+                                    panic!("{:?}::{:?}: Enum should have `variants`&`discriminant_field`, Struct should have `fields`", &cmd.req.cmd, &nested_type.name);
+                                },
+                            }
+                        }).collect()
+                    } else {
+                        vec![]
+                    }
+                };
+
+                // Told you we were going to populate `allowed_extra_types` !
+                {
+                    let mut allowed_extra_types = allowed_extra_types.borrow_mut();
+                    for nested_type in gen_nested_types.iter() {
+                        let name = match nested_type {
+                            GenCmdNestedType::Enum { name, .. } => name,
+                            GenCmdNestedType::Struct { name, .. } => name,
+                        };
+                        allowed_extra_types.insert(name.to_owned(), name.to_owned());
+                    }
+                }
+
+                let gen_req = GenCmdReq {
+                    _cmd: cmd.req.cmd.to_owned(),
+                    kind: match (&cmd.req.unit, &cmd.req.fields) {
+                        (Some(nested_type_name), None) => GenCmdReqKind::Unit {
```

Comment:
This type of struct is called a tuple struct

```rust
// A tuple struct called `Foo`
struct Foo(Bar)
```

Reply:
`Unit` was referring to the fact this thing is defined by a `unit` field in the json.
I don't really care about the name, but I find it more understandable to name this enum after the json spec than after what is generated with it (given the generation step is much more hidden)